reco_utils/recommender/deeprec/graphrec/ranking_metrics.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import time
+import numpy as np
+import pandas as pd
+from reco_utils.common.constants import (
+    DEFAULT_USER_COL,
+    DEFAULT_ITEM_COL,
+    DEFAULT_RATING_COL,
+    DEFAULT_PREDICTION_COL,
+    DEFAULT_K,
+)
+from reco_utils.dataset.pandas_df_utils import (
+    lru_cache_df,
+)
+
+
+@lru_cache_df(maxsize=1)
+def merge_ranking_true_pred(
+    rating_true,
+    rating_pred,
+    col_user,
+    col_item,
+    col_prediction,
+):
+    """
+    Note:
+        Adapted from reco_utils/evaluation/python_evaluation.py.
+        Main difference from original codes: rating_pred has already been processed to show top k items recommended
+            for every user and sorted in descending order by col_prediction.
+    """
+
+    # Make sure the prediction and true data frames have the same set of users
+    common_users = set(rating_true[col_user]).intersection(set(rating_pred[col_user]))
+    rating_true_common = rating_true[rating_true[col_user].isin(common_users)]
+    rating_pred_common = rating_pred[rating_pred[col_user].isin(common_users)]
+    n_users = len(common_users)
+
+    # Return hit items in prediction data frame with ranking information. This is used for calculating NDCG and MAP.
+    # Use first to generate unique ranking values for each item. This is to align with the implementation in
+    # Spark evaluation metrics, where index of each recommended items (the indices are unique to items) is used
+    # to calculate penalized precision of the ordered items.
+    df_hit = rating_pred_common
+    df_hit["rank"] = df_hit.groupby(col_user).cumcount() + 1
+    df_hit = pd.merge(df_hit, rating_true_common, on=[col_user, col_item])[
+        [col_user, col_item, "rank"]
+    ]
+
```

Comment:
Yes, [get_top_k_items in python_evaluation.py](https://github.com/microsoft/recommenders/blob/558733f366983d576953a407ab7180b1642dbc5b/reco_utils/evaluation/python_evaluation.py#L635) takes longer than [get_top_k_scored_items in python_utils.py](https://github.com/microsoft/recommenders/blob/558733f366983d576953a407ab7180b1642dbc5b/reco_utils/common/python_utils.py#L69). The former takes around 78s to get the top k while the latter takes around 16s when I test lightgcn codes with gowalla dataset (not in this repository, it's a large dataset used in the original paper). So I use the latter function in lightgcn.py and modify python_evaluation.py to a new ranking_metrics.py to avoid geting top k items again when calculating the metrics.

Should I add some argument to python_evaluation.py to indicate there is no need to get the top k items?

Reply:
yes, I think you can add some argument to python_evaluation.py to indicate there is no need to get the top k items.  But make sure the new logic would not influence the existing code.

Reply:
+1 to **adding arguments to switch between the options**